pkg/auth/handler/saml/logout.go

```diff
@@ -0,0 +1,238 @@
+package saml
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/lib/saml/samlbinding"
+	"github.com/authgear/authgear-server/pkg/lib/saml/samlerror"
+	"github.com/authgear/authgear-server/pkg/lib/saml/samlprotocol"
+	"github.com/authgear/authgear-server/pkg/lib/saml/samlprotocol/samlprotocolhttp"
+	"github.com/authgear/authgear-server/pkg/util/clock"
+	"github.com/authgear/authgear-server/pkg/util/httproute"
+	"github.com/authgear/authgear-server/pkg/util/log"
+	"github.com/authgear/authgear-server/pkg/util/panicutil"
+)
+
+func ConfigureLogoutRoute(route httproute.Route) httproute.Route {
+	return route.
+		WithMethods("GET", "POST").
+		WithPathPattern("/saml2/logout/:service_provider_id")
+}
+
+type LogoutHandlerLogger struct{ *log.Logger }
+
+func NewLogoutHandlerLogger(lf *log.Factory) *LogoutHandlerLogger {
+	return &LogoutHandlerLogger{lf.New("saml-logout-handler")}
+}
+
+type LogoutHandler struct {
+	Logger      *LogoutHandlerLogger
+	Clock       clock.Clock
+	SAMLConfig  *config.SAMLConfig
+	SAMLService HandlerSAMLService
+
+	ResultWriter SAMLResultWriter
+}
+
+func (h *LogoutHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	serviceProviderId := httproute.GetParam(r, "service_provider_id")
+	sp, ok := h.SAMLConfig.ResolveProvider(serviceProviderId)
+	if !ok {
+		http.NotFound(rw, r)
+		return
+	}
+
+	if !sp.SLOEnabled {
+		http.NotFound(rw, r)
+		return
+	}
+
+	callbackURL := sp.SLOCallbackURL
+	responseBinding := sp.SLOBinding
+
+	defer func() {
+		if err := recover(); err != nil {
+			h.handleUnknownError(rw, r, callbackURL, responseBinding, "", err)
+		}
+	}()
+
+	relayState, result := h.handleSLORequest(r, sp, callbackURL)
+	h.writeResult(rw, r, result, &samlprotocolhttp.WriteOptions{
+		Binding:     responseBinding,
+		CallbackURL: callbackURL,
+		RelayState:  relayState,
+	})
+}
+
+func (h *LogoutHandler) handleSLORequest(
+	r *http.Request,
+	sp *config.SAMLServiceProviderConfig,
+	callbackURL string,
+) (relayState string, result samlprotocolhttp.SAMLResult) {
+	now := h.Logger.Time.UTC()
+	var parseResult samlbinding.SAMLBindingParseResult
+	var logoutRequest *samlprotocol.LogoutRequest
+
+	// Get data with corresponding binding
+	var err error
```

Comment:
1. Prefer more than one methods returning (A, error). The error could be a error result that can be written directly.
2. One method to see the flow e.g. handleSLORequest

```
func handleSLORequest() {
  request, err := parse(r)
}
```


```
var err error

defer func() {
  if err != nil {
    // ...
  }
}()



switch {
case A
  if err != nil {
    return
  }
case B
}
```